read_buffer/src/lib.rs

```diff
@@ -395,7 +389,10 @@ impl Database {
         let names = chunks
             .iter()
             .fold(BTreeSet::new(), |mut names, chunk| {
-                names.append(&mut chunk.table_names(&predicate));
+                // notice that `names` is pushed into the chunk `table_name`
+                // implementation. This ensure we don't process more tables than
+                // we need to.
+                names.append(&mut chunk.table_names(&predicate, &names));
```

Comment:
I wonder if you could arrange for this to look something like

```suggestion
                let names = chunk.table_names(&predicate, names));
```

(aka pass in ownership of names down the stack rather)

I think given how you are using fold here it might look quite nice